crates/datasources/src/common/sink/csv.rs

```diff
@@ -0,0 +1,141 @@
+use std::fmt;
+use std::sync::Arc;
+
+use async_trait::async_trait;
+use datafusion::arrow::csv::{Writer as CsvWriter, WriterBuilder as CsvWriterBuilder};
+use datafusion::arrow::record_batch::RecordBatch;
+use datafusion::common::Result as DfResult;
+use datafusion::error::DataFusionError;
+use datafusion::physical_plan::insert::DataSink;
+use datafusion::physical_plan::SendableRecordBatchStream;
+use futures::StreamExt;
+use object_store::path::Path as ObjectPath;
+use object_store::ObjectStore;
+use tokio::io::{AsyncWrite, AsyncWriteExt};
+
+use crate::common::errors::Result;
+
+use super::SharedBuffer;
+
+const BUFFER_SIZE: usize = 2 * 1024 * 1024;
+
+#[derive(Debug, Clone)]
+pub struct CsvSinkOpts {
+    /// Delimiter between values.
+    pub delim: u8,
+    /// Include header.
+    pub header: bool,
+}
+
+impl Default for CsvSinkOpts {
+    fn default() -> Self {
+        CsvSinkOpts {
+            delim: b',',
+            header: true,
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct CsvSink {
+    store: Arc<dyn ObjectStore>,
+    loc: ObjectPath,
+    opts: CsvSinkOpts,
+}
+
+impl fmt::Display for CsvSink {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "CsvSink({}:{})", self.store, self.loc)
+    }
+}
+
+impl CsvSink {
+    pub fn from_obj_store(
+        store: Arc<dyn ObjectStore>,
+        loc: impl Into<ObjectPath>,
+        opts: CsvSinkOpts,
+    ) -> CsvSink {
+        CsvSink {
+            store,
+            loc: loc.into(),
+            opts,
+        }
+    }
+
+    async fn stream_into_inner(&self, mut stream: SendableRecordBatchStream) -> Result<usize> {
+        let (_id, obj_handle) = self.store.put_multipart(&self.loc).await?;
+        let mut writer = AsyncCsvWriter::new(obj_handle, BUFFER_SIZE, &self.opts);
+
+        while let Some(batch) = stream.next().await {
+            let batch = batch?;
+            writer.write_batch(&batch).await?;
+        }
+        writer.finish().await?;
+
+        Ok(0)
```

Comment:
Should we always be returning 0, or should this be counting the number of rows being written.

Reply:
Aah yes, we should technically be returning number of rows.